spec.emu

```diff
@@ -221,6 +210,90 @@ contributors: Surma, Dan Ehrenberg
+        <emu-table id="table-additional-fields-of-source-text-module-records" caption="Additional Fields of Source Text Module Records" oldids="table-38">
+          <table>
+            <tr>
+              <th>
+                Field Name
+              </th>
+              <th>
+                Value Type
+              </th>
+              <th>
+                Meaning
+              </th>
+            </tr>
+            <tr>
+              <td>
+                [[ECMAScriptCode]]
+              </td>
+              <td></td>
+              <td></td>
+            </tr>
+            <tr>
+              <td>
+                [[Context]]
+              </td>
+              <td></td>
+              <td></td>
+            </tr>
+            <tr>
+              <td>
+                [[ImportMeta]]
+              </td>
+              <td></td>
+              <td></td>
+            </tr>
+            <tr>
+              <td>
+                [[ImportEntries]]
+              </td>
+              <td></td>
+              <td></td>
+            </tr>
+            <tr>
+              <td>
+                [[LocalExportEntries]]
+              </td>
+              <td></td>
+              <td></td>
+            </tr>
+            <tr>
+              <td>
+                [[IndirectExportEntries]]
+              </td>
+              <td></td>
+              <td></td>
+            </tr>
+            <tr>
+              <td>
+                [[StarExportEntries]]
+              </td>
+              <td></td>
+              <td></td>
+            </tr>
+            <tr>
+              <td>
+                <ins>[[TopLevelScriptOrModule]]</ins>
+              </td>
+              <td>
+                <ins>a Script Record, a Module Record, *null*.</ins>
+              </td>
+              <td>
+                <p><ins>The top-level context where this module has been defined. The default value of [[TopLevelScriptOrModule]] is the Module Record itself, representing a Module Record created by the host with a ParseModule call. It can only be a Script Record or *null* when the Module Record is the result of importing a ModuleBlock.</ins></p>
+                <emu-note>
+                  <p><ins>An example of when [[TopLevelScriptOrModule]] can be *null* is in a web browser host. There, if a user clicks on a control given by</ins></p>
+
+                  <pre><code class="html">&lt;button type="button" onclick="import(module {})"&gt;Click me&lt;/button&gt;</code></pre>
+
+                  <p><ins>there will be no active script or module at the time the <emu-xref href="#sec-import-calls">`module {}`</emu-xref> expression runs.</ins></p>
+                </emu-note>
```

Comment:
This is one way to achieve it, but there is another way, and that is to simply impose the invariant in a top down way.

Basically spec language doesn't need to define the data constructs like you would in a programming model, spec language defines invariants.

We can define the invariant quite fine without having this field by just imposing it as an invariant that the resolultion should be the same for all blocks and top-level modules.

This then relaxes the requirements on hosts in maintaining that linkage so that eg the top-level module could be GC'd fine in future while the module block sticks around.

Basically simply a text invariant might suffice for `HostResolveImportedModule` as described in https://github.com/tc39/proposal-js-module-blocks/pull/68/files#r912411086.

Reply:
> This then relaxes the requirements on hosts in maintaining that linkage

Since resolution/evaluation of module blocks is done within ECMA-262 without going through host hooks, I don't think that they can maintain that linkage. But you are right, we can avoid storing the whole Module Record to make GC easier (technically GC is already allowed as long as you replace them with something that "maintains" identity since we never check the contents of the [[TopLevelModuleOrScript]] slot, but yes it's hard with this PR).

Reply:
We should specify that, just like Proxy does - failing to satisfy invariants throws an exception.

Reply:
In this case checking if an invariant is satisfied is as much work for an engine as doing the bookkeeping work itself.

For example, in this case:
```js
// ./main.js

import * as x from "./foo.js";
import * as y from "./foo.js";
```

If you have a loader with `importHook(importerURL: string, specifier: string)`, we can specify two alternatives:
- We call `importHook("./main.js", "./foo.js")` only the first time, and use its result for both imports
- We call `importHook("./main.js", "./foo.js")` twice, and throw if it has two different results

In both cases the engine needs to keep a reference to `importHook("./main.js", "./foo.js")`'s first result, so the second option adds burden on the JS dev (since they need to manually cache `importHook` results) without removing any complexity from the engine.

Reply:
The current situation is that we have an invariant that is defined but cannot be verified by the spec. It's definitely worth considering if we can improve that but certainly not a given either. Perhaps ECMA-262 should just directly handle the resolution caching itself then? That might also help as a way to build the invariant more easily for custom loaders? It's worth noting that import maps in browsers also seem to be heading in this direction (https://github.com/guybedford/import-maps-extensions/issues/17#issuecomment-1150050026). Not sure just asking the questions.

Reply:
Uh well, the current spec text of this proposal handles caching! It was needed to specify how caching works with module blocks. There is a new realm.[[ModuleMap]] list that does exactly that: https://tc39.es/proposal-js-module-blocks/#sec-code-realms.

I proposed adding it directly to ECMA-262, but the editors discussed about it in their weekly call and asked to bring it in with a proposal that needs it (https://github.com/tc39/ecma262/pull/2806#issuecomment-1170571998).